### DISCUSSION TITLE
Q: 自定义服务端，整合前端到现有项目 || Q: Customize the server and integrate the front end into existing projects

### DISCUSSION THREAD

**Discussion by AnzhiZhang** [walinejs/waline#2334] [Q&A] [answered]
### Describe the bug | 问题描述

我注意到提供了服务端的 [API 列表](https://waline.js.org/reference/server/api.html)，我是否可以用现有项目的服务端实现这些接口，并结合自定义用户系统 #1795，实现使用 waline 的前端，将数据存储在现有项目的后端？

### Website URL | 问题网站

.

### Where your waline deploy? | 服务部署在哪里？

Vercel (Default)

### Where your comment data store? | 数据存储在哪里？

PostgreSQL

---
### Describe the bug | Problem description

I noticed that the [API list](https://waline.js.org/reference/server/api.html) is provided on the server side. Can I use the server side of the existing project to implement these interfaces and combine it with custom users? System #1795, Implementing a frontend using waline to store data in the backend of an existing project?

### Website URL | Problem website

.

### Where your waline deploy? | Where is the service deployed?

Vercel (Default)

### Where your comment data store? | Where is the data stored?

PostgreSQL

**Comment by lizheming** [comment, marked as answer]
理论上是可以的，你都自己实现服务端了，其实没必要纠结 #1795 了~ 

---
Theoretically, it's possible. You've implemented the server yourself, so there's no need to worry about #1795~

**Reply by AnzhiZhang** [reply]
又看了一下 #1795，似乎主要是后端的修改，在前端这块应该如何处理呢？比如已登录用户同时将 waline 设为登录

---
After looking at #1795 again, it seems that it is mainly a back-end modification. How should the front-end be handled? For example, a logged-in user also sets waline as login

**Reply by lizheming** [reply]
https://github.com/walinejs/waline/blob/main/packages/client/src/composables/userInfo.ts 存在了 localStorage 里面

---
https://github.com/walinejs/waline/blob/main/packages/client/src/composables/userInfo.ts exists in localStorage

**Reply by AnzhiZhang** [reply]
请问评论的渲染到 HTML 是后端实现的吗？我看接口的返回有 comment 和 orig，是前端直接使用 comment 展示了吗

---
Is the rendering of comments to HTML implemented on the backend? I see that the interface returns comment and orig. Does the front end directly use comment to display?

**Reply by lizheming** [reply]
是的，默认是使用 comment 的，orig 是在管理员已登录的场景下前端需要做内容编辑时使用的。

---
Yes, comment is used by default, and orig is used when the front end needs to edit content when the administrator is logged in.

**Reply by AnzhiZhang** [reply]
您好，我将 Waline 的 Vue 组件接到了现有项目中，但是使用 useStorage 和 useLocalStorage 更新用户信息都无法响应式更新 waline 的用户状态，我还尝试了使用 v-if 来刷新 waline 组件，也没有效果。只有刷新页面才能生效，让 waline 获取新的用户信息，请问可以提供一些解决思路吗？

相关代码：

```html
<ClientOnly>
  <Waline
      v-if="walineReady"
      :serverURL="blogStore.walineServerURL"
      :path="post.permalink"
      :lang="locale"
      :locale="walineLocale"
      :emoji="walineEmoji"
      commentSorting="hottest"
      login="disable"
      :search="false"
  />
</ClientOnly>
```

```ts
import {useLocalStorage} from "@vueuse/core";
const walineReady = ref(false);
if (process.client) {
  const localStorage = useLocalStorage("WALINE_USER", {});
  if (authStore.authenticated) {
    localStorage.value = {
      token: authStore.token,
      type: authStore.isAdmin || post.authorId == authStore.authorization?.accountId! ? "administrator" : "guest",
    };
  } else {
    localStorage.value = {};
  }
  walineReady.value = true;
}
```

PS：进度情况。通过后端自定义 comment 相关接口，前端使用组件配置 + 修改 localStorage，已经基本实现了自定义用户系统和后端。美中不足的有两点：1. 上文提到的数据绑定更新问题；2. waline 的鉴权方式是 header `Authorization: Bearer {token}`，token 来自于本地存储的 token，我的系统正好兼容这个鉴权方式，如果可以允许传入 preRequest 函数，实现其他的鉴权方式就好了，不过这个可能会和 #1795 中提到的长期计划一起实现？

**Reply by lizheming** [reply]
提的问题比较多，建议新开 issue 沟通。1 的问题看的不是很明白，建议具体说下背景和步骤，如果有最小可复现 demo 当然是最好的。2 这个可以新开 discussion 说下你的想法一块聊聊。

---
There are many questions raised, so it is recommended to open a new issue for communication. Question 1 is not very clear. It is recommended to explain the background and steps in detail. It would be best if there is a minimum reproducible demo. 2 You can open a new discussion to share your thoughts and let’s talk together.

**Reply by AnzhiZhang** [reply]
第一个问题已经新开了 #2522，第二个从后端来说有 api 层面的，有数据库层面的，甚至可以通过 SDK 来做，前端也不太好处理，暂时没有完整的想法。

---
The first issue has been opened as #2522. The second one has API level and database level from the back end, and can even be done through SDK. The front end is not easy to handle, so I don’t have a complete idea for the time being.

**Reply by lizheming** [reply]
第二个问题建议还是拆一个新的 discussion 讨论下，主要是我不清楚你的背景和目的诉以及诉求，#1795 这个需求已经实现了，暂时无这块的长期计划哦~
